cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+//! A late-stage redundant-reload remover, that runs after registers have been
+//! allocated and stack slots have been given specific offsets.
+
+use crate::cursor::{Cursor, EncCursor};
+use crate::dominator_tree::DominatorTree;
+use crate::ir::{
+    Ebb, Function, Inst, InstBuilder, InstructionData, Opcode, StackSlotData, Value, ValueLoc,
+};
+use crate::isa::{RegInfo, RegUnit, TargetIsa};
+
+use std::vec::Vec;
+
+/// The remover's state.
+pub struct RedundantReloadRemover {
+    // The running environment used to detect redundancies.  This is reinitialised at the start of
+    // each EBB.
+    env: AvailEnv,
+}
+
+impl RedundantReloadRemover {
+    /// Create a new remover state.
+    pub fn new() -> Self {
+        Self {
+            env: AvailEnv::new(0),
+        }
+    }
+
+    /// Clear the state of the remover.
+    pub fn clear(&mut self) {
```

Comment:
Since `init` is called for each EBB's beginning, and init calls clear, maybe we can:

- inline the content of clear in init (and make it slightly smarter)
- and remove this public function clear, since it has no effect?

(I think a previous comment suggested to do so, sorry for the back and forth)

Going a bit further, I think this should be in `AvailEnv` itself, since it modifies its attributes, so it could be reused by the AvailEnv's ctor too, and maybe called "reset". Then the run function could call this between the different EBBs.